labrad/util/mux.py

```diff
@@ -0,0 +1,57 @@
+from twisted.internet import defer
+
+from labrad import util
+
+
+def after(delay):
+    """Return a Deferred that will fire after the given delay (in seconds).
+
+    This can be useful to implement a "timeout" option when calling select.
+    """
+    return util.wakeupCall(delay)
+
+
+def select(options):
+    """Select from among a set of Deferreds the first one that fires.
+
+    Args:
+        options (Dict[str, Deferred]): Map from key name (str) to Deferred. The
+            key associated with the first Deferred that fires will be returned
+            in the resulting Selection to identify which one fired.
+
+    Returns:
+        (Deferred[Selection]): A Deferred that will fire with a Selection that
+        contains the key of the selected Deferred and the resulting value or
+        error.
+    """
```

Comment:
That won't work for a few reasons. First, the outer method wants to return the `result` Deferred, so here we have to pass a value to it by calling its `.callback` method, not by assigning to it. Also, if we try to assign to `deferred_received` inside `handle`, then python will consider it a local variable so `handle` will fail because the local variable won't be defined yet when it's checked in the `if` clauses. In python 3 you can declare a variable as "nonlocal" to make this work, but in python 2 you'd have to store the shared value inside a mutable data structure, like
```python
deferred_received = [False]
@defer.inlineCallbacks
def handle(key, deferred):
    try:
        value = yield deferred
        if not deferred_received[0]:
            result.callback(Selection(key, value))
            deferred_received[0] = True
    except Exception as e:
        ...
```